### ISSUE TITLE
Unexpected variation in results of `textDocument/declaration` and `textDocument/definition`

### ISSUE THREAD

**Issue by siegel** [clangd/clangd#713]
Please describe the problem.

I'm writing up "Jump to Definition" and "Jump to Declaration" features in my LSP client, and observing some unexpected behavior. Specifically, it appears as though `textDocument/declaration` and `textDocument/definition` _both_ return varying results based on the starting location, even though in objective terms the declaration and definition haven't moved in the source file.

With reference to the enclosed Zip archive of my test workspace: [LSPTest.zip](https://github.com/clangd/clangd/files/6093992/LSPTest.zip)

In `main.mm`, send `textDocument/declaration` for a location in the center of `Mumble ` on line 34.
=> `textDocument/declaration` returns location of the method declaration on line 7 in the class definition. This is as expected.

then, send `textDocument/declaration` for a location in the center of `Mumble ` on line 7.
=> textDocument/declaration returns location of method *definition* on line 16.

What I expected to happen in the second instance is that `textDocument/declaration` would return the same answer as in the first instance, since the method *declaration* of `Mumble` is still on line 7. (And I would argue that returning the location of a *definition* is incorrect, insofar as the request was explicitly for the *declaration*.)

A similar experiment with `textDocument/definition` returns results that I contend are incorrect:

In `main.mm`, send `textDocument/definition` for a location in the center of `Mumble ` on line 34. This is request 
=> `textDocument/definition` returns the location of the method *declaration* on line 7 in the class definition. I think it should return the location of the method *definition* on line 16.

then, send `textDocument/definition` for a location in the center of `Mumble` on line 7.
=> `textDocument/definition` returns the location of the method definition on line 16, which is what I expected (and what I believe should have happened in the first instance).

Hopefully this makes sense; but please let me know if I can provide any additional details or diagnostics.

**Logs**
Excerpt from verbose log attached.
[LanguageServerProtocol-Objective-C++.log](https://github.com/clangd/clangd/files/6093617/LanguageServerProtocol-Objective-C%2B%2B.log)

**System information**
Output of `clangd --version`: `clangd version 11.1.0`
Editor/LSP plugin: n/a
Operating system: macOS 10.15.7

**Comment by HighCommander4** [comment]
Based on experience with other (non-LSP) editors, the "go to definition" command toggling between the definition and declaration of a function/method is a common behaviour that users are familiar with.

As a user, having to reach for a different shortcut (to invoke a different command) to go from definition --> declaration than from declaration --> definition would be extra overhead that I would consider unfortunate.

**Comment by sam-mccall** [comment]
Right, this is deliberate - while textDocument/declaration exists, it's pretty obscure and in practice we've mostly seen people happy if "go to definition" is the magic "do what I mean" button.

We could probably provide a way to turn this off either for editors (protocol level) or users (config).

(There's a fair argument this is a client concern, but the protocol doesn't really seem geared for that, and we've mostly seen clients to be quite "thin" in practice)

**Comment by siegel** [comment]
Thanks for your comments!

The problem I'm having is that I believe the decisions of where to go should be left to the client, not made by the server. As the client developer I am uniquely qualified to make an informed decision based on the responses received from the server, *if* those responses are consistent and reflect a reasonable (and reasonably literal) interpretation of the LSP spec.

In context of the languages supported by `clangd`, a given symbol may have a declaration, a definition, both, or neither. In some cases, the definition and declaration can be in the same location. So I think it would be entirely appropriate for `clangd` to return answers that reflects these realities, and not make something up.

I freely acknowledge that in many places LSP is vague or leaves out important details about what behaviors are expected of the client and/or server in response to various requests. This became painfully clear to me some time ago and I'm doing my best to work with it.

But even so, regarding the current issue, I feel very strongly that if I send `textDocument/declaration` to the server, it should respond with the location(s) of the *declaration* of that symbol, never the definition. Conversely, when the server receives `textDocument/definition`, it should return the *definition* of the symbol, never the declaration.

If the symbol is declared and defined in the same location, or if either declaration or definition are absent, then so be it — I can cope with that on the client side because I expect it. And if the symbol declaration (or definition) happens to include (or be on the same line as) the location I supplied in the request, I can cope with that as well.

In other words, I think the "DWIM" behavior should be the province of the client, not the server. I am perfectly willing to do the work required to DWIM, even if it's something that other client developers would rather not deal with. :-)

Having said all that, clearly this behavior is intentional in `clangd`, and changing it would break existing clients, and I respect that. So I will implement special cases to work with it, and hope that it never changes.

However, I strongly advocate for `clangd` to implement rigorous behavior for `textDocument/declaration` and `textDocument/definition`, *if* the client advertises in its capabilities at `initialize` time that it wants that behavior. (I'm certain it shouldn't be user configurable.) I don't know offhand whether the protocol spec provides for server-specific capability requests, if not I have some half-baked thoughts that hopefully would be feasible for `clangd` without breaking the rules of the spec.

Thanks for listening.

**Comment by mclow** [comment]
> Based on experience with other (non-LSP) editors, the "go to definition" command toggling between the definition and declaration of a function/method is a common behaviour that users are familiar with.
> 
> As a user, having to reach for a different shortcut (to invoke a different command) to go from definition --> declaration than from declaration --> definition would be extra mental overhead that I would consider unfortunate.

(Echoing @sam-mccall)
Agree completely, *BUT* that's not the province of clangd. That's for the editor to implement.
clangd is there to provide services to LSP editors (and other clients), not to users.

**Comment by sam-mccall** [comment]
> that's not the province of clangd. That's for the editor to implement.

I'd personally prefer if the protocol and ecosystem worked this way.
But in practice, the protocol is treated as a list of features by clients.
Methods correspond directly to affordances in the VSCode UI, and other editors follow suit. I'm not aware of any clients that implement such features.

I'm happy if BBEdit is to be one! Let's by all means add a strictness capability to change this behavior. But just dropping it is not going to server the majority of users well. I'd suggest `textDocument.definition.strict` in the ClientCapabilities to request this behavior. (And declaration)

(I think there are going to be some awkward limits to trying to build semantic features client-side on top of LSP. It doesn't tend to carry rich enough semantics, IMO because VSCode doesn't use them and that's their top priority)

**Comment by asmwarrior** [comment]
Any progress about the `textDocument.definition.strict`? In our clangd client([Re: Code completion using LSP and clangd](https://forums.codeblocks.org/index.php/topic,24357.msg172188.html#msg172188)), we do need such feature to distinguish `goto definition` and `goto declaration`. Thanks.

**Comment by HighCommander4** [comment]
I'm happy to provide guidance if someone would like to have a go at implementing `textDocument.definition.strict`.
